server/src/main/java/org/elasticsearch/common/time/DateUtilsRounding.java

```diff
@@ -0,0 +1,181 @@
+        long i2 = (utcMillis >> 1) + APPROX_MILLIS_AT_EPOCH_DIVIDED_BY_TWO;
+        if (i2 < 0) {
+            i2 = i2 - unitMillis + 1;
+        }
+        int year = (int) (i2 / unitMillis);
+
+        long yearStart = utcMillisAtStartOfYear(year);
+        long diff = utcMillis - yearStart;
+
+        if (diff < 0) {
+            year--;
+        } else if (diff >= MILLIS_PER_DAY * 365L) {
+            // One year may need to be added to fix estimate.
+            long oneYear;
+            if (isLeapYear(year)) {
+                oneYear = MILLIS_PER_DAY * 366L;
+            } else {
+                oneYear = MILLIS_PER_DAY * 365L;
+            }
+
+            yearStart += oneYear;
+
+            if (yearStart <= utcMillis) {
+                // Didn't go too far, so actually add one year.
+                year++;
+            }
+        }
+
+        return year;
+    }
+
+    // see org.joda.time.chrono.BasicGJChronology
+    static int getMonthOfYear(final long utcMillis, final int year) {
+        // Perform a binary search to get the month. To make it go even faster,
+        // compare using ints instead of longs. The number of milliseconds per
+        // year exceeds the limit of a 32-bit int's capacity, so divide by
+        // 1024. No precision is lost (except time of day) since the number of
+        // milliseconds per day contains 1024 as a factor. After the division,
+        // the instant isn't measured in milliseconds, but in units of
+        // (128/125)seconds.
+
+        int i = (int)((utcMillis - utcMillisAtStartOfYear(year)) >> 10);
+
+        // There are 86400000 milliseconds per day, but divided by 1024 is
+        // 84375. There are 84375 (128/125)seconds per day.
+
+        return
+            (isLeapYear(year))
+                ? ((i < 182 * 84375)
+                ? ((i < 91 * 84375)
+                ? ((i < 31 * 84375) ? 1 : (i < 60 * 84375) ? 2 : 3)
+                : ((i < 121 * 84375) ? 4 : (i < 152 * 84375) ? 5 : 6))
+                : ((i < 274 * 84375)
+                ? ((i < 213 * 84375) ? 7 : (i < 244 * 84375) ? 8 : 9)
+                : ((i < 305 * 84375) ? 10 : (i < 335 * 84375) ? 11 : 12)))
+                : ((i < 181 * 84375)
+                ? ((i < 90 * 84375)
+                ? ((i < 31 * 84375) ? 1 : (i < 59 * 84375) ? 2 : 3)
+                : ((i < 120 * 84375) ? 4 : (i < 151 * 84375) ? 5 : 6))
+                : ((i < 273 * 84375)
+                ? ((i < 212 * 84375) ? 7 : (i < 243 * 84375) ? 8 : 9)
+                : ((i < 304 * 84375) ? 10 : (i < 334 * 84375) ? 11 : 12)));
+    }
+
+    // see org.joda.time.chrono.BasicGJChronology
+    static long getTotalMillisByYearMonth(final int year, final int month) {
+        if (isLeapYear(year)) {
+            return MAX_TOTAL_MILLIS_BY_MONTH_ARRAY[month - 1];
+        } else {
+            return MIN_TOTAL_MILLIS_BY_MONTH_ARRAY[month - 1];
+        }
+    }
+
+    private static final long[] MIN_TOTAL_MILLIS_BY_MONTH_ARRAY;
```

Comment:
Should we move the static fields and the static initializer to the top of the class?